feat(API): Add optional projectId parameter to credential creation

hi120ki opened the pull request:
## Summary

<!--
Describe what the PR does and how to test.
Photos and videos are recommended.
-->

Previously, [credentials created via the Public API](https://docs.n8n.io/api/api-reference/#tag/credential/post/credentials) were always saved to the user's personal project.
This change adds an optional projectId parameter to the POST /credentials endpoint, allowing credentials to be created in any project where the user has credential:create permission.

The implementation follows [the pattern used in the REST API credential creation endpoint](https://github.com/n8n-io/n8n/blob/a2efb1fe7e0a8963ac68f82d96a4027ad77b0598/packages/cli/src/credentials/credentials.service.ts#L397-L454).
If projectId is not provided, the credential is saved to the user's personal project, maintaining backward compatibility.

## Related Linear tickets, Github issues, and Community forum posts

<!--
Include links to **Linear ticket** or Github issue or Community forum post.
Important in order to close *automatically* and provide context to reviewers.
https://linear.app/n8n/issue/
-->
<!-- Use "closes #<issue-number>", "fixes #<issue-number>", or "resolves #<issue-number>" to automatically close issues when the PR is merged. -->


## Review / Merge checklist

- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) <!--
   **Remember, the title automatically goes into the changelog.
   Use `(no-changelog)` otherwise.**
-->
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [x] Tests included. <!--
   A bug is not considered fixed, unless a test is added to prevent it from happening again.
   A feature is not complete without tests.
-->
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Expose optional projectId in credential creation to create credentials in a specified project with permission checks, updating services, OpenAPI, and tests.
> 
> - **Public API**:
>   - `POST /credentials` now accepts optional `projectId`; handler forwards it to `saveCredential` and returns proper `403`/`404` on permission/not-found.
>   - OpenAPI updated: request/response schemas include `projectId`; responses add `403` and `404`.
> - **Services**:
>   - Credential creation flows switched to `ProjectService.getProjectForCredentialCreation` to resolve target project and enforce `credential:create` permissions; removed direct repository checks.
> - **Tests**:
>   - Added tests for `saveCredential` with/without `projectId`.
>   - Added tests for `ProjectService.getProjectForCredentialCreation` (success, forbidden, not found, optional transaction).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 512e5d4a5154fc8141da1141c4810e04cfb42e02. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

hi120ki committed:
feat(API): Add optional projectId parameter to credential creation

nikhilkuria commented:
Hi, thanks for making this contribution. Can you please add unit tests for the changes?

hi120ki committed:
Merge branch 'master' into feat/public-api-credential-project-id

hi120ki committed:
Add: 2 unit tests

Signed-off-by: Hi120ki <12624257+hi120ki@users.noreply.github.com>

nikhilkuria commented on packages/cli/src/public-api/v1/handlers/credentials/credentials.service.ts:
```diff
@@ -62,15 +65,32 @@ export async function saveCredential(
-		);
+		const project =
+			projectId === undefined
+				? await projectRepository.getPersonalProjectForUserOrFail(user.id, transactionManager)
```
I think this is a valid comment. Can you please extract a common method to ProjectService

hi120ki replied:
@nikhilkuria Hello! I added `getProjectForCredentialCreation` on ProjectService to unify the project permission checks. Can you please take a look?

hi120ki committed:
Merge branch 'master' into feat/public-api-credential-project-id

hi120ki committed:
Add: getProjectForCredentialCreation to ProjectService

Signed-off-by: Hi120ki <12624257+hi120ki@users.noreply.github.com>

hi120ki commented on packages/cli/src/services/project.service.ee.ts:
```diff
@@ -627,4 +627,41 @@ export class ProjectService {
+		if (typeof projectId === 'string' && project === null) {
+			throw new ForbiddenError(
+				"You don't have the permissions to save the credential in this project.",
+			);
```
While I understand the concern about the misleading error message, keeping the current behavior is intentional for security reasons. Returning a NotFoundError when the project doesn’t exist would allow attackers to infer which project IDs are valid through enumeration or probing. By always returning a ForbiddenError, we prevent this kind of project ID search attack and avoid unintentionally exposing information about the existence of projects to unauthorized users.

nikhilkuria reviewed (dismissed):
Nice work! Thanks a lot for working on this.

nikhilkuria committed:
Merge branch 'master' into feat/public-api-credential-project-id

hi120ki committed:
Fix: pnpm typecheck

Error: src/credentials/__tests__/credentials.service.test.ts(71,3): error TS2554: Expected 12 arguments, but got 13.
Signed-off-by: Hi120ki <12624257+hi120ki@users.noreply.github.com>

hi120ki committed:
Fix: does not create the credential in a specific project if the user is not part of the project

Signed-off-by: Hi120ki <12624257+hi120ki@users.noreply.github.com>

nikhilkuria committed:
Merge branch 'master' into feat/public-api-credential-project-id

hi120ki commented:
@nikhilkuria The upstream was changed https://github.com/n8n-io/n8n/pull/22355 to refactor the `packages/cli/src/credentials/credentials.service.ts`'s save function.
Can you please help me to achieve the goal?
